Add st.experimental_(get|set)_query_params capability

zhaoooyue opened the pull request:
**Issue:**  Issue #798 #302 #1098 This PR is proposing a simple change to add capability to set and get current query strings from browser URL in user's browser tab (without triggering a navigation event!). By introducing this feature to `st.experimental` namespace, a Streamlit user can easily bookmark an application using query strings and reopen the application else where and render all states using the query strings!

**Demo**
![experimental_qs](https://user-images.githubusercontent.com/57884188/79066973-6cf0c300-7cee-11ea-8ee0-ec381b800959.gif)
**Feature:**
:one: User can call `st.experimental_set_query_params` => _to set query params for current browser tab_

Example

- If user’s current browser tab is http://localhost:3000 and user called `st.experimental_set_query_params( {"checkbox1":"true", "multiselect2": 2} )`, the user’s browser tab will become http://localhost:3000/?checkbox1=true&multiselect2=2 and this will NOT cause a redirect action.

:two: User can call `st.experimental_get_query_params` => _to get the current query strings as a dict _{"checkbox1": "true", "multiselect2=2"}_
**Example Usage from above demo**
```
import streamlit as st

st.title("Experimental: Play with Query Strings!")

app_state = st.experimental_get_query_params()
app_state = {k: v[0] if isinstance(v, list) else v for k, v in app_state.items()} # fetch the first item in each query string as we don't have multiple values for each query string key in this example

# [1] Checkbox
default_checkbox = eval(app_state["checkbox"]) if "checkbox" in app_state else False
checkbox1 = st.checkbox("Are you really happy today?", key="checkbox1", value = default_checkbox)
app_state["checkbox"] = checkbox1
st.experimental_set_query_params(**app_state)
st.markdown("")


# [2] Radio
radio_list = ['Eat', 'Sleep', 'Both']
default_radio = int(app_state["radio"]) if "radio" in app_state else 0
genre = st.radio("What are you doing at home during quarantine?", radio_list, index = default_radio)
if genre:
    app_state["radio"] = radio_list.index(genre)
    st.experimental_set_query_params(**app_state)
st.markdown("")


# [3] Text Input
default_title = app_state["movie"] if "movie" in app_state else ""
title = st.text_input('Movie title', value = default_title)
app_state["movie"] = title
st.experimental_set_query_params(**app_state)

```

**Contribution License Agreement**

By submiting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

tvst commented:
Hey @zhaoooyue , just (belatedly) following up from our meeting: we'll have a meeting on our end about this proposal and will get back to you.

Thanks for sending the PR!

zhaoooyue commented:
@tvst Thanks Thiago! let me know if any further changes are needed, i.e. tests.

k4r1 commented:
Maybe `get_query_string` and `set_query_string` might be a bit better for the name? That way we don't get in the way of any future plans to allow for different routes within the application?

orieger commented:
Hey streamlit Team, thanks for this marvelous application - helps so much.
And thank you @zhaoooyue for this useful additional feature!
One question though: Is the following behaviour intended? Seems like you can't call the app with query params and then retrieve these params with `st.get_url()`.

![streamlit_get_url](https://user-images.githubusercontent.com/10261781/77534598-4252dd80-6e99-11ea-9d4c-3f7dba8e58c2.gif)

Code is
```
import streamlit as st
url = st.get_url()
url
```

orieger commented:
Hey streamlit Team, thanks for this marvelous application - helps so much.
And thank you @zhaoooyue for this useful additional feature!
One question though: Is the following behaviour intended? Seems like you can't call the app with query params and then retrieve these params with st.get_url().

![streamlit_get_url](https://user-images.githubusercontent.com/10261781/77534799-9d84d000-6e99-11ea-8c1e-d2a3e9f4c834.gif)

Code:
```
import streamlit as st
url = st.get_url()
url
```

treuille commented:
Hey Karl and @zhaoooyue:

We hope this email finds you well and safe during this unprecedented time!

### How we're thinking about experimental features in general

We appreciate your innovative proposal and it prompted a lot of discussion! 🙏🏻 

We want to make sure that we're responsive to the needs and efforts of the community. We are currently planning to release a new `st.experimental` package ([link](https://github.com/streamlit/streamlit/pull/1240) - more details coming soon) in order to create a faster and more predictable path for new features and ideas to become part of Streamlit. We also ask that groups keep in mind our [contribution guidelines](https://github.com/streamlit/streamlit/wiki/Contributing). We would greatly appreciate your feedback on these guidelines and on this new process for experimental features.

### Specific thoughts on this proposal

In thinking about this proposal our goals were to:

1. Help unblock your team. (We really appreciate Oak North's support for Streamlit.)
2. Rationalize this idea with our [roadmap](https://discuss.streamlit.io/t/the-streamlit-roadmap-big-plans-for-2020/2054), and in particular with:
    1. Our [plans for state](https://discuss.streamlit.io/t/programmable-state-for-streamlit/2052) (which somewhat conflict with your proposal)
    2. A general proposal which has been floating around to encode widget state in URL parameters. 

### Specific recommendations for this PR

To combine these goals we kindly and respectfully ask that you:

1. Add `st.set_url` and `st.get_url` to the new `st.experimental` package.

This will also allow us to solicit more feedback from the community about this proposal.

We apologize for the delay getting back to you on this. We greatly appreciate Oak North's support for Streamlit! ❤️

zhaoooyue commented:
> Hey streamlit Team, thanks for this marvelous application - helps so much.
> And thank you @zhaoooyue for this useful additional feature!
> One question though: Is the following behaviour intended? Seems like you can't call the app with query params and then retrieve these params with st.get_url().
> 
> ![streamlit_get_url](https://user-images.githubusercontent.com/10261781/77534799-9d84d000-6e99-11ea-8c1e-d2a3e9f4c834.gif)
> 
> Code:
> 
> ```
> import streamlit as st
> url = st.get_url()
> url
> ```

Hi @orieger - yes this behaviour is expected as the `st.get_url` can only retrieve the url which is modified by `st.set_url` . In that sense, it is not capable of reading arbitrary url which are not set by `st.set_url`. 

As you can see this is currently an experimental feature, and the design behind this is to not block other routes to be introduced by future Streamlit features. This is also the reason why the current design is not causing a redirect in the browser. 

However I certainly feel the need of your requirement - similar to @tvst mentioned earlier
>  _.....when a user loads a Streamlit URL, the app would also automatically check the params to set up its widgetState before making a rerun request to the server..._ 

These are two different routes to handle this new capability - let's wait Streamlit team's reply on this on a possible design which is also in line with future features 😄

zhaoooyue commented:
@treuille Thanks for looking at the PR! 

As requested, I have move the two new APIs to `st.experimental` namespace and updated the description to include the updated code snippets to utilise this feature.

You were also mentioning that there is somewhat conflicts between this PR and the team's Plan for [State](https://discuss.streamlit.io/t/programmable-state-for-streamlit/2052) . Appreciate it if more info/feedback can be provided such I can incorporate this into the `plan`, or even re-vamp it! 

Not last - thanks for this amazing tool, as always 💯

tvst commented:
Hey, apologies for the delay in reviewing! Will look at this today.

tvst commented on frontend/src/App.tsx:
```diff
@@ -454,6 +460,14 @@ export class App extends PureComponent<Props, State> {
+   * @param newUrl string
+   */
+  handleNewUrlChange = (newUrl: string): void => {
+    window.history.pushState({}, "", newUrl ? newUrl : "/")
```
There's likely a bug here: the URL should be prefixed with the `basePath` from `lib.UriUtil.getWindowBaseUriParts()`, since people can host apps under non-root paths too.

And following up on item 1 from the root comment in this PR, when you change the code to only allow setting the query string, then the `protocol`, `host`, and `basePath` should all come from `getWindowBaseUriParts`.

zhaoooyue replied:
this has been changed to append queryString using `window.history.pushState` such the browser will not be redirected or refreshed. 

this will only be called upon user calls `st.experimental.set_query_string(some_value)`. on server side I have added validation to ensure frontend will only receive valid query strings, so we should be good here. Proof: [Click me](https://github.com/zhaoooyue/streamlit/blob/69f0673d6900bac0d6a30f07f62114ebbd5a6d23/lib/streamlit/__init__.py#L625-L630)

tvst commented on frontend/src/App.tsx:
```diff
@@ -339,6 +340,11 @@ export class App extends PureComponent<Props, State> {
+    // send current URL to server
+    const backMsg = new BackMsg({ urlInfo: window.location.href })
+    backMsg.type = "urlInfo"
+    this.sendBackMsg(backMsg)
```
Nit: move this into its own method, so you can call it here like this:

```js
this.sendCurrentUrlToServer()
```

This way the code is more self-documenting.


Also: please add a comment inside that method saying "This is only used in st.experimental.get_url"

tvst commented on lib/streamlit/ReportSession.py:
```diff
@@ -116,6 +116,8 @@ def __init__(
 
         self._scriptrunner = None
+        self.base_url = None
+        self.query_url = ""
```
See item 2 in the top-level comment on this code review. I think we shouldn't add these here for an experimental feature, especially since the `get_url` and `set_url` functions are fairly "prototype quality" at this point. Instead, let's just have `get_url` and `set_url` monkey-patch the `ReportSession` object.

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -102,6 +102,8 @@
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
 from streamlit.errors import StreamlitAPIException
 from streamlit.proto import BlockPath_pb2 as _BlockPath_pb2
 from streamlit.util import functools_wraps as _functools_wraps
```
Nit: Sort imports.

tvst commented on proto/streamlit/proto/BackMsg.proto:
```diff
@@ -49,5 +49,8 @@ message BackMsg {
     bool close_connection = 10;
+
+    // Request to pass current URL to backend
+    string url_info = 11;
```
This name should start with a verb, to follow the convention in the file.

Also, this gets filled with the full URL, right? If so, a better name would be just `store_url`. Or actually `store_query_string`, to conform to item 1 of the root-level comment.

(I considered `set_url` but that would make it seem like this field is used to change the URL that is set in the browser. Instead, this is the exact opposite: this field is used to tell the server which URL is currently set in the browser. So `store_url` seemed a little better — although I don't love it either :wink: )

zhaoooyue replied:
I have used the `store_url` for now (sorry for the weird name 😉 )

Rational: the only BackMsg being sent to Server is upon user first opens the tab. It would make sense the sent the browser URL back to Server. Server will do the parsing and store the query strings only (such we don't need to parse on frontend)

tvst commented on proto/streamlit/proto/ForwardMsg.proto:
```diff
@@ -72,6 +72,9 @@ message ForwardMsg {
     string ref_hash = 11;
+
+    // new_url for curious users to set params
+    string new_url = 12;
```
* Move this above the "Other messages" separator.
* Rename to `set_url` (actually `set_query_string`, so it's clearer that this will actually change the URL that is shown in the browser location bar
* Mention in the comment that the URL is only sent from the frontend to the backend when the tab first loads. After that, the backend assumes it knows the URL.
* Mention something like: "this is only used in st.experimental.set_url. Do not write code that depends on this!" since experimental features should be easily removable.
* I don't really understand the comment "new_url for curious users to set params". What do you mean by "curious users"? A better comment would be something like:

```proto
// Tells the browser to update the query string in the URL shown in its address bar.
// This is only used in st.experimental.set_query_string. Do not write code that depends on this!
string set_query_string = 12;
```

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
+def _get_url():
+    """
+    this should just return the url that the current session is running on
+    """
```
Use [Numpy](https://numpydoc.readthedocs.io/en/latest/format.html) and [PEP257 format](https://www.python.org/dev/peps/pep-0257/) for docstring. In particular:
* The first line should start at the `"""`
* Docs should follow normal rules of capitalization, punctuation
* Docs should describe arguments and return value

For example

```py
def _get_url():
    """Return the URL that is currently showing in the browser URL bar.

    Returns
    -------
    str
      The entire URL, from the protocol (e.g. "http://") to the end of the query
      string.
    """
```

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
+                return session.base_url + str(session.query_url)
+
+
+def _set_url(url_to_set):
```
Based on item 1 from the root comment in this PR, this should now be `_set_query_string(query_dict)`, where `query_dict` is a dict that will get internally converted to a query string. See [urllib](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode)

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
+def _set_url(url_to_set):
+    """
+    this should just append/replace url_to_set to the current session's existing hostname.
+    """
```
Use [Numpy](https://numpydoc.readthedocs.io/en/latest/format.html) and [PEP257 format](https://www.python.org/dev/peps/pep-0257/) for docstring.

tvst commented on proto/streamlit/proto/BackMsg.proto:
```diff
@@ -49,5 +49,8 @@ message BackMsg {
     bool close_connection = 10;
+
+    // Request to pass current URL to backend
+    string url_info = 11;
```
Mention in the comment that the URL is only sent from the frontend to the backend when the tab first loads. After that, the backend assumes it knows the URL.

Also mention something like: "this is only used in st.experimental.get_query_string. Do not write code that depends on this!" since experimental features should be easily removable.

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
+    """
+    this should just return the url that the current session is running on
+    """
+    ctx = _get_report_ctx()
```
As mentioned in item 2 of this PR's root-level comment, this part of the PR is OK for now but only because this is an experimental feature. So it's OK if it's "prototype-quality".

But if we were to bring this feature into Streamlit's core `st` namespace we'd have to do a deep rethinking here.

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
+            if not session.base_url:
+                raise StreamlitAPIException(
+                    "Something is wrong when fetching current browser URL. Please try again."
+                )
```
A better phrasing would be "Error reading the app's URL from the browser. Please try again."

tvst commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +582,38 @@ def _transparent_write(*args):
     return args
 
 
+def _get_url():
```
Based on item 1 from the root comment in this PR, this should be renamed `_get_query_string`, and it should return the query string.

It would also be great if the return value were a dict rather than a str. So if the query string is "?foo=123&bar=42&greetings=hello&greetings=hi" then the dict would be:

```
{
  foo: 123,
  bar: 42,
  greetings: ["hello", "hi"],
}
```

See [urllib.parse](https://docs.python.org/3/library/urllib.parse.html)

zhaoooyue commented:
thanks @tvst ! I will update the PR to address your comments accordingly

tvst commented:
Hey @zhaoooyue , LMK when you need another review.

zhaoooyue commented:
Hi @tvst - thanks for the thorough review lat time! I have updated the PR addressing your comments (except `tests` in progress). Hopefully it has now jumped to `experimental-quality` from `prototype-quality`.  The changes include
* Change to set/get query string everywhere with `urllib` `parse_qs `and `urlencode` . Now the call to `st.experimental.(get|set)_query_string` will operate on `dict` as you suggested. 
* No more modification on `ReportSession` and `ReportContext`. query strings are attached via monkey-patch approach.
* Update comments/docstring based on streamline standard (including explicitly stating `this is used in st.experimental.get_query_string` )
* Add retry logic in `get_query_string` (thanks @orieger for finding the issue)

I have also updated the PR description with this changes as well as a demo on how to bookmark an Streamlit app with query strings (tag #302 and [this](https://github.com/CodeForPhilly/chime/issues/146)).   

@tvst Another review will be greatly appreciated! I will read about Streamlit unit/e2e testing approaches and add tests meanwhile.

monchier commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +584,72 @@ def _transparent_write(*args):
+
+
+def _set_query_string(query_string_dict):
+    """Set the query string in currently browser URL bar and update the query string value in current ReportSession.
```
`current browser`

zhaoooyue replied:
👌

monchier commented on lib/streamlit/server/Server.py:
```diff
@@ -401,8 +402,8 @@ def _loop_coroutine(self, on_started=None):
                 """
-Please report this bug at https://github.com/streamlit/streamlit/issues.
-"""
+                Please report this bug at https://github.com/streamlit/streamlit/issues.
+                """
```
Has `info` automatic dedent?

zhaoooyue replied:
yeah it's somehow decent in current develop branch

monchier commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +584,72 @@ def _transparent_write(*args):
+                        return session.query_string
+                except StreamlitAPIException as e:
+                    if i >= _MAX_RETRIES - 1:
+                        raise e
```
Can you add a log message?

zhaoooyue replied:
will do!

monchier commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +584,72 @@ def _transparent_write(*args):
+    session_infos = Server.get_current()._session_info_by_id
+    _MAX_RETRIES = 3
+    _RETRY_WAIT_SECS = 0.2
+    for socket_handler, session_info in session_infos.items():
+        if session_info.session.enqueue == ctx.enqueue:
+            session = session_info.session
+            # Add retry in case query string message are not processed
+            for i in range(_MAX_RETRIES):
+                try:
+                    if (
+                        not hasattr(session, "query_string")
+                        or session.query_string is None
+                    ):
+                        raise StreamlitAPIException(
+                            "Error reading the current app's query strings from the browser. Please try again."
+                        )
+                    else:
+                        return session.query_string
+                except StreamlitAPIException as e:
+                    if i >= _MAX_RETRIES - 1:
+                        raise e
```
Is this retry loop because the browser may not have sent the message with the query string yet?

zhaoooyue replied:
Yes. The retry loop will make sure the query string are received.

monchier commented on frontend/src/App.tsx:
```diff
@@ -339,6 +341,16 @@ export class App extends PureComponent<Props, State> {
     this.handleSessionStateChanged(sessionState)
+
+    // send current URL to server
+    this.sendCurrentUrlToServer()
```
Does this happens every time query params are set "manually" in the browser?

zhaoooyue replied:
Nope, this only happens when the Streamlit app first loads. Every time query params are set "manually" in the browser -> this method will not be invoked.

monchier commented on frontend/src/App.tsx:
```diff
@@ -339,6 +341,16 @@ export class App extends PureComponent<Props, State> {
+    this.sendCurrentUrlToServer()
+  }
+
+  // This is only used in st.experimental.get_query_string
```
This message is sent in any case, right? Whether we are using the `st.get/set_query...` or not.

zhaoooyue replied:
Yes, I will modify the comment to make it more proper.

monchier commented on lib/streamlit/__init__.py:
```diff
@@ -580,6 +584,72 @@ def _transparent_write(*args):
+    session_infos = Server.get_current()._session_info_by_id
+    _MAX_RETRIES = 3
+    _RETRY_WAIT_SECS = 0.2
+    for socket_handler, session_info in session_infos.items():
+        if session_info.session.enqueue == ctx.enqueue:
+            session = session_info.session
+            # Add retry in case query string message are not processed
+            for i in range(_MAX_RETRIES):
+                try:
+                    if (
+                        not hasattr(session, "query_string")
+                        or session.query_string is None
+                    ):
+                        raise StreamlitAPIException(
+                            "Error reading the current app's query strings from the browser. Please try again."
+                        )
+                    else:
+                        return session.query_string
+                except StreamlitAPIException as e:
+                    if i >= _MAX_RETRIES - 1:
+                        raise e
```
If this is handling a race condition we should likely handle this differently.

zhaoooyue replied:
It is handling a race condition to be honest, any idea how should this be handled better? Appreciate any thoughts on this

zhaoooyue commented:
@monchier Thanks for the review! I will add tests this weekend. Sorry it's delayed a bit due to the current challenging situation (schools are closed 😢 ). 

On where to store the query string - last time @tvst suggested _a monkey-patch ReportSession and store the query string as dict._ Would love to see any design that doesn't rely on ReportSession - however I do find it challenging a bit given `st.experimental.set_query_string` is invoked on the backend and the information needs to be sent to frontend. I'm assuming widgetStatesMgr are doing the same thing to propagate message to backend without out-of-sync issue as well?

Will come back on this once tests are added! Thanks for the review and do stay safe 🙏

monchier commented:
> @monchier Thanks for the review! I will add tests this weekend. Sorry it's delayed a bit due to the current challenging situation (schools are closed 😢 ).

Yeah... same boat here...

> 
> On where to store the query string - last time @tvst suggested _a monkey-patch ReportSession and store the query string as dict._ Would love to see any design that doesn't rely on ReportSession - however I do find it challenging a bit given `st.experimental.set_query_string` is invoked on the backend and the information needs to be sent to frontend. I'm assuming widgetStatesMgr are doing the same thing to propagate message to backend without out-of-sync issue as well?

After clarification, this looks good to me.
> 
> Will come back on this once tests are added! Thanks for the review and do stay safe 🙏

zhaoooyue commented:
Hi @monchier - I have added e2e tests to test `experimental_get|set_query_string`. 

Given the nature of monkey-patch ReportSession to store the query strings, the two new API won't work if it's evaluated without a browser connected (before the get|set messages can communicate). Such that in e2e I am testing the behaviours via a button click.

Let me know your thoughts on this, we could possibly figure out a new approach to overcome this limitation ( or we can assume it's okay for current version to reside in experimental world 😄 )

julien-c commented:
I know it's bad form but just commenting to add a massive `:+1:` on this PR/feature from the Hugging Face team :)

vhoulbreque commented:
Hi @zhaoooyue, this would really be a game-changing feature for my use case. Can I be of any help for the implementation of the last steps of your MR? :)

glebkrapivin commented:
@zhaoooyue Hey, tried to build from your pull requess. Unfortunately, i cannot access 
```st.experimental_get_query_string()``` immediately when the page loads showing message NOT_INITIALIZED. In my case it works only after a trigger some action on the page after it has loaded (like hitting 'r' to rerun). Is there any way to fix this? 
Thanks!

lmeyerov commented:
Continuing from https://discuss.streamlit.io/t/multiple-apps-on-same-server/4216/2

-- This would help us in a near-term workaround for multi-page apps. (We'd still want subpaths long-term for SEO, but for users sharing links, this is great.)

-- Our use case is actually the same kind of dashboard experience, so tying form controls to url parameters is what our users would when sharing configurations of a page, and is standard to many dashboard systems. This is a good building block for getting UI controls to auto-expose this.

-- RE:Polish, As this is being user-exposed, it may make sense to namespace to avoid collisions and prefer st-core in argument order

lmeyerov commented:
@monchier Is this in experimental / any guidance on using it there / thoughts on getting it there?

We were trying on `63.0` via  `streamlit.experimental.get_...`, `streamlit.experimental_get_...` and no go.

tvst commented:
Hey, sorry for the insane delay here. One of our engineers went on paternity leave and this fell through the cracks. 

I'm taking a look at this right now. Will do whatever is left to do to get it merged ASAP

zhaoooyue commented:
Thanks @tvst ! Glad to see the demand behind this PR.

I think currently the major flaw is that given we are monkey-patching `ReportSession` to store the query strings, the two new API won't work if it's evaluated without a browser connected (before the get|set messages can communicate).  So if one user tries to open an app via `http://localhost:3000/?app1=True` **for the first time** -> it will fail to get the query string. We possibly need a more generic approach here. Would appreciate your insights here!

tvst commented:
So since this PR has waited long enough in this queue, I spent half of today working on it so it's mergeable with Streamlit's main branch (i.e. `develop`). See https://github.com/streamlit/streamlit/pull/1743

@zhaoooyue , I'd like to make sure this goes under your name when merged, so you get due credit 😃  . So once my PR is approved, would you mind merging my branch into yours, so I can then merge yours into `develop`? In fact, I'd also really appreciate it if you could read through my PR and see if you have any concerns.

zhaoooyue commented:
@tvst  thanks for the great work on 1743! I will rebase from your branch and test the feature locally again. Will voice out if I see any concerns 😁

tvst commented:
@zhaoooyue , PR #1743 has been reviewed and passes all tests. Feel free to rebase anytime, then let me know when done so I can merge.

WOOHOO!!!!

Thanks again for all your work on this 🙏

tvst commented:
BTW, if you want I can try to push my changes to your branch directly. This usually works, but depends on the access controls settings for your repo

zhaoooyue commented:
@tvst I have rebased from #1743 and upstream develop branch + squashed the commits. Have also verified the behaviour is working fine locally. Example in this PR description is updated accordingly as well. Guess it's now in a good shape to be merged! 😉

benlindsay commented:
This is fantastic! When can I expect this to be included in a release to pypi?

randyzwitch commented:
Hi @benlindsay - 

This should come out in the next minor release, 0.65. But in the meantime, we publish streamlit-nightly, and you can use one of these versions to test the functionality out.

https://pypi.org/project/streamlit-nightly/#history

Best,
Randy

benlindsay commented:
Thanks @randyzwitch! Didn't know about the nightly release, I'll give that a shot

benlindsay commented:
Can someone point me to a session state implementation that's compatible with the latest nightly release? (0.64.1.dev20200731 as of now) I tried [this one](https://gist.github.com/tvst/0899a5cdc9f0467f7622750896e6bd7f) and got this error:
```
ModuleNotFoundError: No module named 'streamlit.server.Server'
Traceback:

File "/----------/streamlit-tabs/venv/lib/python3.8/site-packages/streamlit/script_runner.py", line 324, in _run_script
    exec(code, module.__dict__)
File "/----------/streamlit-tabs/app.py", line 2, in <module>
    import st_state_patch
File "/----------/streamlit-tabs/st_state_patch.py", line 54, in <module>
    from streamlit.server.Server import Server
```

I also tried [this one](https://gist.github.com/tvst/036da038ab3e999a64497f42de966a92) and got this error:
```
ModuleNotFoundError: No module named 'streamlit.ReportThread'
Traceback:

File "/-----------/streamlit-tabs/venv/lib/python3.8/site-packages/streamlit/script_runner.py", line 324, in _run_script
    exec(code, module.__dict__)
File "/-----------/streamlit-tabs/app.py", line 3, in <module>
    import SessionState
File "/-----------/streamlit-tabs/SessionState.py", line 18, in <module>
    import streamlit.ReportThread as ReportThread
```

vhoulbreque commented:
They have been renamed to `streamlit.server.server` and `streamlit.report_thread`.

benlindsay commented:
Thanks @vinzeebreak! I'll give that a shot.

Update: making those changes works. thanks!

jaystary commented:
Was this feature removed again in the latest nightly?
Aside from that i noticed some issues. If i changed and saved my script after the second time, the query parameters would be none although they still are present in the URL.

Also .experimental_set_query_params(dict) wouldnt  allow to take parameters...

This happend with dev20200801

benlindsay commented:
> Also .experimental_set_query_params(dict) wouldnt allow to take parameters...

Is it supposed to take a dict as an argument? For me passing **dict instead of dict works

jaystary commented:
I think something broke with the latest nightly because the entire feature is gone.
Also- the change from streamlit.server.Server to streamlit.server.server seems to have reverted

randyzwitch commented:
@benlindsay or @jaystary can you split this off into a new issue? I want to make sure this  doesn't get lost if people aren't checking a closed PR

benlindsay commented:
I'm not sure it should be a new issue. I just tried with the dev20200801 and the feature is still there, and streamlit.server.server did not revert. @jaystary, my guess is that you installed the nightly version in one environment, then ran streamlit from another environment, which is a super easy mistake to make. I did something similar within the past week or two. Maybe run `which pip` and `which streamlit`, make sure those executables are in the same environment, run `pip list` and verify you have the version of streamlit that you think you have, then run streamlit?

karriebear commented:
For completeness, when running the demo, I noticed a bit of a hiccup.
![ezgif com-optimize](https://user-images.githubusercontent.com/24946400/89989087-6ba8b600-dc4e-11ea-9775-6329ff0d680f.gif)

This is because of an off-by-one error, which arises from Streamlit's execution model (i.e. no callbacks) and cannot be addressed until we make some changes to that.

A breakdown of what's happening:
1. User runs the app for the first time. "Eat" is selected by default.
2. User clicks on "Sleep". This reruns the script from top to bottom.
3. During that rerun, when the execution reaches this line
```app_state = st.experimental_get_query_params()```
the query params in the URL bar are still set to "Eat" (because the code that sets to "Sleep" hasn't executed yet).
4. This means that default_radio will be set to "Eat" and the code that draws the radio buttons will re-set the selected button to "Eat".
```
# default_radio is set back to "Eat"
# even though you clicked on "Sleep" before.
default_radio = int(app_state["radio"]) if "radio" in app_state else 0
genre = st.radio("What are you doing at home during quarantine?", radio_list, index = default_radio)
# genre is set to "Sleep", though.
```
But, interestingly, that code also checks the widget state that came from the browser and returns "Sleep" due to that. So now genre is "Sleep".
5. Now we reach the lines below
```
app_state["radio"] = radio_list.index(genre)
st.experimental_set_query_params(**app_state)
```
At this point the query params are set with radio=2 ("Sleep"). Leading to an inconsistent state!

karriebear commented:
Looks like using @tvst's gist for [SessionState](https://gist.github.com/tvst/036da038ab3e999a64497f42de966a92) will resolve the funny behavior. Here's an updated demo:

```
import streamlit as st
import SessionState

query_params = st.experimental_get_query_params()
app_state = st.experimental_get_query_params()

first_query_params = session_state.first_query_params
session_state = SessionState.get(first_query_params=query_params)

radio_list = ['Eat', 'Sleep', 'Both']

# The trick here is you can't change the default index based on query params on every run!
# The *only time* you do that is on the *first* run.
default_index = eval(first_query_params["radio"][0]) if "radio" in app_state else 0

genre = st.radio("What are you doing at home during quarantine?", radio_list, index=default_index)
app_state["radio"] = radio_list.index(genre)

st.experimental_set_query_params(**app_state)
```

mappingvermont commented:
This is awesome!! Thanks so much @zhaoooyue @tvst @karriebear 👏 👏 👏

WilberDelbrison commented:
Hi All, thanks for this incredible library!
This function to get the parameters is based on session information? I'm using my App with multiple users, and with each new session, I'm not able to read the parameters with st.experimental_get_query_params().
Thanks in advance

karriebear commented:
@WilberDelbrison looking through the PR quickly, it mentions using a client state so it should be based on per user session. If this is still an issue, could you please open a new issue with a reproducible code sample for us to look at?

vini-2907 commented:
> 
> 
> **Issue:** Issue #798 #302 #1098 This PR is proposing a simple change to add capability to set and get current query strings from browser URL in user's browser tab (without triggering a navigation event!). By introducing this feature to `st.experimental` namespace, a Streamlit user can easily bookmark an application using query strings and reopen the application else where and render all states using the query strings!
> 
> **Demo**
> ![experimental_qs](https://user-images.githubusercontent.com/57884188/79066973-6cf0c300-7cee-11ea-8ee0-ec381b800959.gif)
> **Feature:**
> 1️⃣ User can call `st.experimental_set_query_params` => _to set query params for current browser tab_
> 
> Example
> 
>     * If user’s current browser tab is http://localhost:3000 and user called `st.experimental_set_query_params( checkbox1=True, multiselect2=2 )`, the user’s browser tab will become http://localhost:3000/?checkbox1=true&multiselect2=2 and this will NOT cause a redirect action.
> 
> 
> 2️⃣ User can call `st.experimental_get_query_params` => _to get the current query strings as a dict _{"checkbox1": "true", "multiselect2=2"}_
> **Example Usage from above demo**
> 
> ```
> import streamlit as st
> 
> st.title("Experimental: Play with Query Strings!")
> 
> app_state = st.experimental_get_query_params()
> app_state = {k: v[0] if isinstance(v, list) else v for k, v in app_state.items()} # fetch the first item in each query string as we don't have multiple values for each query string key in this example
> 
> # [1] Checkbox
> default_checkbox = eval(app_state["checkbox"]) if "checkbox" in app_state else False
> checkbox1 = st.checkbox("Are you really happy today?", key="checkbox1", value = default_checkbox)
> app_state["checkbox"] = checkbox1
> st.experimental_set_query_params(**app_state)
> st.markdown("")
> 
> 
> # [2] Radio
> radio_list = ['Eat', 'Sleep', 'Both']
> default_radio = int(app_state["radio"]) if "radio" in app_state else 0
> genre = st.radio("What are you doing at home during quarantine?", radio_list, index = default_radio)
> if genre:
>     app_state["radio"] = radio_list.index(genre)
>     st.experimental_set_query_params(**app_state)
> st.markdown("")
> 
> 
> # [3] Text Input
> default_title = app_state["movie"] if "movie" in app_state else ""
> title = st.text_input('Movie title', value = default_title)
> app_state["movie"] = title
> st.experimental_set_query_params(**app_state)
> ```
> 
> **Contribution License Agreement**
> 
> By submiting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

can You pls help me do this how to do it how to install am very new to streamlit with clear steps it would be more helpful
how can I install st.experimental

karriebear commented:
@vini-2907 there's no need to install anything else. Anything under the experimental namespace are available in streamlit just prefaced with `experimental_`. You should be able to run the code sample as is. If you need additional help, please head over to our [forums](https://discuss.streamlit.io/c/Questions/5). We have a great community ready to help!

You can learn more about our experimental namespace in our [docs](https://docs.streamlit.io/en/stable/pre_release_features.html?highlight=experimental#experimental)

hfwittmann commented:
> Looks like using @tvst's gist for [SessionState](https://gist.github.com/tvst/036da038ab3e999a64497f42de966a92) will resolve the funny behavior. Here's an updated demo:
> 
> ```
> import streamlit as st
> import SessionState
> 
> query_params = st.experimental_get_query_params()
> app_state = st.experimental_get_query_params()
> 
> first_query_params = session_state.first_query_params
> session_state = SessionState.get(first_query_params=query_params)
> 
> radio_list = ['Eat', 'Sleep', 'Both']
> 
> # The trick here is you can't change the default index based on query params on every run!
> # The *only time* you do that is on the *first* run.
> default_index = eval(first_query_params["radio"][0]) if "radio" in app_state else 0
> 
> genre = st.radio("What are you doing at home during quarantine?", radio_list, index=default_index)
> app_state["radio"] = radio_list.index(genre)
> 
> st.experimental_set_query_params(**app_state)
> ```

Works for me except the order of two lines has to be changed like so:

from this:
 ```
first_query_params = session_state.first_query_params
session_state = SessionState.get(first_query_params=query_params)
 ```

to that:
 ```
session_state = SessionState.get(first_query_params=query_params)
first_query_params = session_state.first_query_params
 ```

And for completeness' sake: for the SessionState I had to use https://gist.github.com/FranzDiebold/898396a6be785d9b5ca6f3706ef9b0bc
which is an updated version referenced in @tvst's gist for SessionState

quantoid commented:
@karriebear thanks for the explanation (above) but I don't understand why the default from the URL parameters is being used once the dashboard has already loaded. The [API docs](https://docs.streamlit.io/en/stable/api.html#streamlit.radio) suggest that the default `index` is only used when the page is loaded, and not when redrawn in response to widget changes:

> **index** (int) – The index of the preselected option _on first render_.

Shouldn't all widgets always show and return any value that comes from the browser, and only use the given default if none?

franekp commented:
Hi Everyone,

Thanks for adding these powerful capabilities to Streamlit! Building on your work, I made a library of url-aware versions of input controls that automatically sync their state with the url query string: [streamlit-permalink](https://github.com/franekp/streamlit-permalink/).

Hopefully some day these ideas will find their way into Streamlit core and such external libraries would no longer be needed :)
